server-spi-private/src/main/java/org/keycloak/protocol/oidc/grants/OAuth2GrantManager.java

```diff
@@ -0,0 +1,110 @@
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.protocol.oidc.grants;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import org.keycloak.OAuth2Constants;
+import org.keycloak.common.util.MultivaluedHashMap;
+import org.keycloak.events.EventType;
+
+/**
+ * A class to register and resolve OAuth 2.0 Grant Type implementations
+ *
+ * @author <a href="mailto:demetrio@carretti.pro">Dmitry Telegin</a>
+ */
+public class OAuth2GrantManager {
+
+    private static final MultivaluedHashMap<String, OAuth2GrantType> GRANT_TYPE_MAP = new MultivaluedHashMap<>();
+    private static final Map<String, EventType> EVENT_TYPE_MAP = new HashMap<>();
+
+    static {
+            EVENT_TYPE_MAP.put(OAuth2Constants.AUTHORIZATION_CODE, EventType.CODE_TO_TOKEN);
+            EVENT_TYPE_MAP.put(OAuth2Constants.REFRESH_TOKEN, EventType.REFRESH_TOKEN);
+            EVENT_TYPE_MAP.put(OAuth2Constants.PASSWORD, EventType.LOGIN);
+            EVENT_TYPE_MAP.put(OAuth2Constants.CLIENT_CREDENTIALS, EventType.CLIENT_LOGIN);
+            EVENT_TYPE_MAP.put(OAuth2Constants.TOKEN_EXCHANGE_GRANT_TYPE, EventType.TOKEN_EXCHANGE);
+            EVENT_TYPE_MAP.put(OAuth2Constants.UMA_GRANT_TYPE, EventType.PERMISSION_TOKEN);
+            EVENT_TYPE_MAP.put(OAuth2Constants.CIBA_GRANT_TYPE, EventType.AUTHREQID_TO_TOKEN);
+            EVENT_TYPE_MAP.put(OAuth2Constants.DEVICE_CODE_GRANT_TYPE, EventType.OAUTH2_DEVICE_CODE_TO_TOKEN);
+    }
+
+    /**
+     * Register OAuth 2.0 grant type
+     * @param grant grant type to register
+     */
+    public static void register(OAuth2GrantType grant) {
+        GRANT_TYPE_MAP.add(grant.getGrantType(), grant);
+    }
+
+    /**
+     * Resolve grant type implementation from grant type and request context
+     * @param grantType grant type
+     * @param context grant request context
+     * @return grant type implementation
+     */
+    public static Optional<OAuth2GrantType> resolve(String grantType, OAuth2GrantType.Context context) {
+        return resolve(resolve(grantType), context);
+    }
+
+    /**
+     * Return registered implementations for the given grant type
+     * @param grantType grant type
+     * @return list of implementations
+     */
+    public static List<OAuth2GrantType> resolve(String grantType) {
+        return GRANT_TYPE_MAP.get(grantType);
```

Comment:
The point of having `GRANT_TYPE_MAP` is to ensure performance and robustness with the growing number of Grant Type implementations.

As far as I understand, our plan is to get rid of grant type specific SPIs (Token Exchange SPI, Assertion Grant SPI etc.; please correct me @pedroigor if I'm wrong) and to migrate to one unified Grant Type SPI instead. This will already result in 7 grant types + 3 Token Exchange variants, at the very least. Considering the number of upcoming specifications that rely on Token Exchange and new grant types, one day we can easily end up with 15+ built-in grants, plus customers adding 1-2 on top of that. In such case, cycling over `supports()` for _every_ token request will be an overkill. Moreover, one single poorly designed implementation could slow down or even break the whole token issuance process, DoSing the server. This is why I suggested to filter the candidates using grant type name (string), then refine the selection using `supports()`.

As for potential hot-reload, I don't think this should be problematic since factories will register themselves in the map during loading (`postInit()`) and de-register when unloading (`close()`), so the map is fully dynamic.

But I agree that the PR is already complex. For now, we can use a simplified approach that you've suggested, and then add two-phase resolution as a follow-up, WDYT?

Reply:
I think using `id` for grant type would be nice. This means that we don't need `supports()` method, which could mean further simplification?

I am not sure if there are some corner cases (like for example if there is some specification supporting something like "dynamic grant" , which means that `grant_type` can be any parameter in the format of like `urn:foo:bar:(.*)` . In that case `id` won't work and we would need `supports()` ? But if existing grants don't need that, we can perhaps just rely on id?